mountpoint-s3/src/data_cache/disk_data_cache.rs

```diff
@@ -246,6 +234,97 @@ impl DataCache for DiskDataCache {
                 _ => Err(DataCacheError::InvalidBlockContent),
             };
         };
+        Ok(file.stream_position()? as usize)
+    }
+
+    fn is_limit_exceeded(&self, size: usize) -> bool {
+        match self.limit {
+            CacheLimit::Unbounded => false,
+            CacheLimit::TotalSize { max_size } => size > max_size,
+            CacheLimit::AvailableSpace { min_ratio } => {
+                let stats = match fs2::statvfs(&self.cache_directory) {
+                    Ok(stats) => stats,
+                    Err(error) => {
+                        warn!(?error, "unable to determine available space");
+                        return false;
+                    }
+                };
+                (stats.available_space() as f64) < min_ratio * (stats.total_space() as f64)
```

Comment:
I think we should also treat `total_space` == 0 as "unable to determine available space" too.